src/Microsoft.Health.Fhir.Core/Features/Operations/DataConvert/DataConvertEngine.cs

```diff
@@ -0,0 +1,150 @@
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.DataConvert
+{
+    public class DataConvertEngine : IDataConvertEngine
+    {
+        private readonly IContainerRegistryTokenProvider _containerRegistryTokenProvider;
+        private readonly ITemplateProviderFactory _templateProviderFactory;
+        private readonly ILogger<DataConvertEngine> _logger;
+
+        private readonly Dictionary<string, IFhirConverter> _dataConverterMap = new Dictionary<string, IFhirConverter>();
+        private const char ImageRegistryDelimiter = '/';
+        private const string DefaultTemplateReference = "microsofthealth/fhirconverter:default";
+
+        public DataConvertEngine(
+            IContainerRegistryTokenProvider containerRegistryTokenProvider,
+            ITemplateProviderFactory templateProviderFactory,
+            ILogger<DataConvertEngine> logger)
+        {
+            EnsureArg.IsNotNull(containerRegistryTokenProvider, nameof(containerRegistryTokenProvider));
+            EnsureArg.IsNotNull(templateProviderFactory, nameof(templateProviderFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _containerRegistryTokenProvider = containerRegistryTokenProvider;
+            _templateProviderFactory = templateProviderFactory;
+            _logger = logger;
+
+            _dataConverterMap.Add(DataConvertInputDataType.Hl7v2.ToString(), new Hl7v2Processor());
+        }
+
+        public async Task<DataConvertResponse> Process(DataConvertRequest convertRequest, CancellationToken cancellationToken)
+        {
+            // We have embedded a default template set in the templatemanagement package.
+            // If the template set is the default refence, we don't need to retrieve token.
+            var accessToken = string.Empty;
+            if (!IsDefaultTemplateReference(convertRequest.TemplateSetReference))
+            {
+                _logger.LogInformation("Using a custom template set for data conversion.");
+                var registryServer = ExtractRegistryServer(convertRequest.TemplateSetReference);
+                accessToken = await _containerRegistryTokenProvider.GetTokenAsync(registryServer, cancellationToken);
+            }
+            else
+            {
+                _logger.LogInformation("Using the default template set for data conversion.");
+            }
+
+            // Fetch templates
+            ITemplateProvider templateProvider;
+            try
+            {
+                templateProvider = await _templateProviderFactory.CreateAsync(convertRequest.TemplateSetReference, accessToken, cancellationToken);
+            }
+            catch (ContainerRegistryAuthException authEx)
+            {
+                _logger.LogError(authEx, "Failed to access container registry: unauthorized.");
+                throw new GetTemplateSetFailedException(string.Format(Resources.GetTemplateSetFailed, authEx.Message), authEx);
```

Comment:
>GetTemplateSetFailedException [](start = 26, length = 29)

This plethora of exceptions would be thrown, and caught up in `OperationOutcomeExceptionFilterAttribute`, but since there is no changes in that filter, they all endup with BadRequest outcome.
Reading log errors I wouldn't say they all related to user providing wrong data.

Reply:
Agree. We'll handle the exceptions caused in service side in the filter.

Reply:
I have updated the service exceptions in the exceptionfilter :)